### DISCUSSION TITLE
Web application for Bisq 2

### DISCUSSION THREAD

**Discussion by chimp1984** [bisq-network/bisq2#350] [General]
For supporting self hosted node projects (e.g. Umbrel, BtcPay Server,...) we should provide a web frontend based on the REST Api.
@salokiam has started to work on that. First goal is to identify the API endpoints for realistic use cases and implement the functionality of the basic use cases.

We should think about which web framework would fit best also in context of potential re-use for a light mobile version (iOS).

**Comment by salokiam** [comment]
Thnking about a web-framework for having a frontend for the self-hosted servers, here are some feature we should take into account for selecting such a framework:

- should fit to our technology stack, so it should be a Java framework
- we are developing UIs with a lot of functionality, so we need a lot of prebuild components which look good out of the box, and have easy binding to our java objects (validating, bean access,...)
- we develop UIs using MVC, so it should support these kind of patterns.
- reduced set of libraries and dependencies (no spring)
- enable modern web front-end (of course)
- easy enough for java developers to learn, as we dont want to deep dive into html/javascript programming

If you have any other relevent points to put on the list, please comment.

**Reply by salokiam** [reply]
One of the framework coming immediatly to my mind is [vaadin](https://vaadin.com/flow), which I have been uing lately  a lot.

**Reply by ghost** [reply]
I have experience in creating single page applications in [VueJS](https://vuejs.org/examples/#modal). It's not Java based but it's learning curve is fairly flat. I could optionally prepare small POC (app that consumes 2-3 bisq2 endpoints with look&feel of bisq2).

**Reply by ripcurlx** [reply]
> If you have any other relevent points to put on the list, please comment.

The only danger is with using Vaadin IMO, that we need to take care not to bypass a REST API, directly accessing services via the framework. Otherwise we'll end up with a crippled API no one else can use.

**Reply by salokiam** [reply]
With Vaadin, we could but dont need to use the REST API as backend. There are some projects I know, they do use their own REST API to access the backend. Right now I had in mind for the REST API only the use case of trading bots, but we could also look into the use case of having the REST API for any alternative UI to be build on. 
In the scenario, where the user installls a bisq2 instance on umbrel (or any other node), we could have an UI for mobile devices, which access the (private) bisq2 instance on umbrel. That would probably be an use case of the REST API. On the other hand having a web-ui on umbrel directly, a mobil user could connect to his umbrel instance using usual tor protocol.
If the web-ui needs to use the REST API, then it would have an impact on the complexity as the REST API would need to be developed in parallel which needs more time.

**Comment by EmmanuelEklipse** [comment]
How would this project be hosted to ensure that it stays decentralized? Surely hosting on a centralized server compromises the very essence of it?

**Reply by apemithrandir** [reply]
The above suggestions relate to a setup where you have your own server running at home and then connect to it remotely via a web front end. Like you would with self-hosted Kodi or Nextcloud.

**Comment by HenrikJannsen** [comment]
There have been a prototype been implemented but the project did not got continued as Bisq Easy does not require the maker to be online and that would have been a major use case to run Bisq headlessly. 
For Musig it would still make sense but it is unrealistic that we can provide the dev resources for that, thus I close the discussion.
